ambry-rest/src/main/java/com/github/ambry/rest/ServerSecurityChecker.java

```diff
@@ -0,0 +1,88 @@
+import com.github.ambry.server.ServerSecurityService;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SslHandler;
+import java.util.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that performs any security validation on the new HTTP2 connection.
+ */
+@ChannelHandler.Sharable
+public class ServerSecurityChecker extends ChannelInboundHandlerAdapter {
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final ServerSecurityService serverSecurityService;
+  private final ServerMetrics serverMetrics;
+
+  public ServerSecurityChecker(ServerSecurityService serverSecurityService, ServerMetrics serverMetrics) {
+    this.serverSecurityService =
+        Objects.requireNonNull(serverSecurityService, "server security service can not be null");
+    this.serverMetrics = Objects.requireNonNull(serverMetrics, "serverMetrics can not be null");
+  }
+
+  @Override
+  public void channelActive(ChannelHandlerContext ctx) throws Exception {
+    logger.trace("Channel Active " + ctx.channel().remoteAddress());
+    validateSslConnection(ctx);
+    super.channelActive(ctx);
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+    logger.trace("Channel Inactive " + ctx.channel().remoteAddress());
+    super.channelInactive(ctx);
+  }
+
+  /**
+   * Listen to the handshake future for the SSL termination, then apply the security policies, potentialy closing the
+   * connection in case security check fails.
+   * @param ctx the {@link ChannelHandlerContext}.
+   */
+  private void validateSslConnection(ChannelHandlerContext ctx) throws Exception {
+    SslHandler sslHandler = ctx.pipeline().get(SslHandler.class);
+    if (sslHandler != null) {
+      sslHandler.handshakeFuture().addListener(future -> {
+        if (!future.isSuccess()) {
+          logger.debug("SSL handshake failed for channel: {}", ctx.channel(), future.cause());
+        } else {
+          logger.debug("SSL handshake succedded for channel: {}", ctx.channel(), future.cause());
+          try {
+            serverSecurityService.validateConnection(ctx, (r, e) -> {
+              if (e != null) {
+                logger.error("security validation failed for channel: {}", ctx.channel(), e);
+                serverMetrics.serverValidateConnectionFailure.inc();
+                ctx.channel().close();
+              } else {
+                logger.info("security validation succeeded for channel: {}", ctx.channel());
+                serverMetrics.serverValidateConnectionSuccess.inc();
+              }
+            });
+          } catch (Exception e) {
+            logger.error("security validation failed for channel: {}", ctx.channel(), e);
+            serverMetrics.serverValidateConnectionFailure.inc();
+            ctx.channel().close();
```

Comment:
what If there is stream channel has been establish?

Reply:
I verified that stream channel is not established at this point. Since this code is getting executed in the same nioEventLoopGroup thread, server will not be processing any other HTTP2 frames for this connection before this handshake callback finishes.